### DISCUSSION TITLE
CNN in dlib - mmod_human_face_detector

### DISCUSSION THREAD

**Discussion by ntd1683** [davisking/dlib#2904] [Q&A] [answered]
Hello, I'm a student,
I want to learn about how cnn works in dlib when using cnn_face_detection_model_v1 mmod_human_face_detector

Can you help me, or read its processing flow ?

Thank you very much

**Comment by arrufat** [comment]
Sure, check this article: https://blog.dlib.net/2017/08/
Even though it's for cars, it works the same as the one trained on faces.

**Comment by ntd1683** [comment]
Is there any way I can read the mod file mmod_human_face_detector.dat

**Comment by arrufat** [comment]
I don't know what you mean by "read" the file, just use it as in the example:
- C++: http://dlib.net/dnn_mmod_face_detection_ex.cpp.html
- Python: http://dlib.net/cnn_face_detector.py.html

If you want to inspect the weights, instead, you can use `dlib::net_to_xml`.
Here's a blog post that does that to visualize the learned convolutional filters:
- https://jacobgil.github.io/deeplearning/hallucinating_faces_dlib_pytorch

If you want to actually see the model architecture, you can use `dlib::net_to_dot`.

![mmod_human_face_detector](https://github.com/davisking/dlib/assets/1671644/f96d0dc5-df4d-42b0-b90d-1dde0beac9fc)

**Reply by ntd1683** [reply]
Thank you very much , This is exactly what I'm looking for

**Reply by ntd1683** [reply]
Can you tell me where you found this photo? I want to know so I can cite the source in the report

**Reply by arrufat** [reply, marked as answer]
This image was generated using the `dlib::net_to_dot` function in the C++ example I mentioned above.

```c++
    net_type net;
    deserialize(argv[1]) >> net;  
    net_to_dot(net, "mmod_human_face_detector.dot");
```

Then:

```console
$ dot -T png < mmod_human_face_detector.dot > mmod_human_face_detector.png
```

You can also generate a PDF:

```console
$ dot -T pdf < mmod_human_face_detector.dot > mmod_human_face_detector.pdf
```

You must have [Graphviz](https://www.graphviz.org/) installed for this to work (the `dot` executable is part of Graphviz).

Here's the generated file, so you can modify it and re-run the `dot` command on it: [mmod_human_face_detector.dot](https://github.com/davisking/dlib/files/13820129/mmod_human_face_detector.txt)

For the citation, just [cite dlib](http://dlib.net/faq.html#HowcanIcitedlib).

**Reply by ntd1683** [reply]
thank you very much
